### ISSUE TITLE
twisted.conch.test.test_ssh.SSHProtocolTestCase.testOurServerOurClient encounters numerous real errors which ignores

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4490]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4490|
|Type|defect|
|Created|2010-06-05 12:56:37Z|
|Branch|https://github.com/twisted/twisted/tree/channel-open-error-4490|

If the two `flushErrors` calls are removed from the code invoked by this test case, the test fails like this:

```
===============================================================================
[ERROR]: twisted.conch.test.test_ssh.SSHProtocolTestCase.testOurServerOurClient

Traceback (most recent call last):
Failure: twisted.conch.error.ConchError: (3, 'unknown channel')
===============================================================================
[ERROR]: twisted.conch.test.test_ssh.SSHProtocolTestCase.testOurServerOurClient

Traceback (most recent call last):
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/log.py", line 84, in callWithLogger
    return callWithContext({"system": lp}, func, *args, **kw)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/log.py", line 69, in callWithContext
    return context.call({ILogContext: newCtx}, func, *args, **kw)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/context.py", line 59, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/context.py", line 37, in callWithContext
    return func(*args,**kw)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/conch/ssh/service.py", line 44, in packetReceived
    return f(packet)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/conch/ssh/connection.py", line 140, in ssh_CHANNEL_OPEN
    common.NS(textualInfo) + common.NS(''))
exceptions.TypeError: unsupported operand type(s) for &: 'str' and 'long'
===============================================================================
[ERROR]: twisted.conch.test.test_ssh.SSHProtocolTestCase.testOurServerOurClient

Traceback (most recent call last):
  File "/home/exarkun/Projects/Twisted/trunk/twisted/conch/ssh/session.py", line 67, in request_exec
    self.session.execCommand(pp, f)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/conch/test/test_ssh.py", line 114, in execCommand
    raise error.ConchError('bad exec')
twisted.conch.error.ConchError: ('bad exec', None)
-------------------------------------------------------------------------------
```

Two of these may be expected, but the middle one certainly isn't and should cause the test to fail (and then should be fixed).

<details><summary>Searchable metadata</summary>

```
trac-id__4490 4490
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_channel_open_error_4490 branches/channel-open-error-4490
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1275742597000000 1275742597000000
changetime__1275963065000000 1275963065000000
version__None None
owner__ 
cc__ivank
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [29200]) Branching to 'channel-open-error-4490'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun removed owner|
|-|-|

See also [#3657](https://github.com/twisted/twisted/issues/3657) and [#3071](https://github.com/twisted/twisted/issues/3071).  I've worked around the problem in the branch with a minimal change, but eventually `ConchError` should really be fixed, like for real.

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|ivank set owner to @exarkun|
|-|-|

1. I'm very worried about `if isinstance(textualInfo, int):`. I looked around to see if `reasonCode` could be a `long`, and it looks like it's possible. Can you change it check for (int, long), or check the type of the other variable instead?

2. Needs a news file. If there is a user-visible bugfix here (even an error getting logged more correctly), that should be mentioned.

3. In `test_ssh.py`, could you add a third argument to `self.assertEquals(len(errors), 2)` to print `errors` if the assertion fails?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [29233]) Be more informative if a different number of errors than expected shows up

refs [#4490](https://github.com/twisted/twisted/issues/4490)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [29235]) Correct the comment about which channel causes an error to be logged

refs [#4490](https://github.com/twisted/twisted/issues/4490)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [29237]) Add a couple unity tests that exercise the exception handling behavior in ssh_CHANNEL_OPEN

refs [#4490](https://github.com/twisted/twisted/issues/4490)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [29238]) Also handle errors with codes represented as longs

refs [#4490](https://github.com/twisted/twisted/issues/4490)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun removed owner|
|-|-|

Thanks for the review.  I've addressed all the points, I think.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/channel-open-error-4490)

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|ivank set owner to @exarkun|
|-|-|

Thanks for those new unit tests.

1. Please add another line of vertical whitespace at the end of `test_connection.TestAvatar.lookupChannel`.

2. Please remove one line of vertical whitespace at the end of `test_ssh.SSHProtocolTestCase.testOurServerOurClient`

3. I'm having a lot of trouble parsing

```
              # Some bytes identifying the request and including the error code
              # (7b in hex) and the error message.
```

Perhaps the solution is to change "and including" to ", including" ?

Everything else looks good, +1 to merging if those are addressed.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [29254]) Fix whitespace issues; reword the comment in the new test_connection helper method; swap a failIf for an assertFalse, rename testOurServerOurClient and expand its docstring slightly

refs [#4490](https://github.com/twisted/twisted/issues/4490)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set status to `closed`|
|-|-|

(In [29255]) Merge channel-open-error-4490

Author: exarkun
Reviewer: ivank
Fixes: [#4490](https://github.com/twisted/twisted/issues/4490)
Refs: [#3071](https://github.com/twisted/twisted/issues/3071)
Refs: [#3657](https://github.com/twisted/twisted/issues/3657)

Handle `ConchError` instances which are misconstructed (by having their arguments
reversed) in the Conch SSH channel open code.  This turns a failed channel open
into a protocol-compliant error response, whereas before it resulted in a lost
connection.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
